src/tests/JIT/opt/AssertionPropagation/ArrBoundMinLength.cs

```diff
@@ -0,0 +1,395 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+
+class Program
+{
+    private static int returnCode = 100;
+
+    private static int[] arr = new int[6];
+
+    public static int Main(string[] args)
+    {
+        RunTestThrows(Tests.GreaterOutOfBound);
+        RunTestThrows(Tests.GreaterEqualOutOfBound);
+        RunTestThrows(Tests.LessOutOfBound);
+        RunTestThrows(Tests.LessEqualsOutOfBound);
+        RunTestThrows(Tests.EqualsOutOfBound);
+        RunTestThrows(Tests.EqualsReversedOutOfBound);
+        RunTestThrows(Tests.NotEqualsOutOfBound);
+        RunTestThrows(Tests.NegativeIndexesThrows);
+        RunTestThrows(TestsEarlyReturn.GreaterOutOfBound);
+        RunTestThrows(TestsEarlyReturn.GreaterEqualOutOfBound);
+        RunTestThrows(TestsEarlyReturn.LessOutOfBound);
+        RunTestThrows(TestsEarlyReturn.LessEqualsOutOfBound);
+        RunTestThrows(TestsEarlyReturn.NotEqualsOutOfBound);
+        RunTestThrows(TestsEarlyReturn.EqualsOutOfBound);
+
+        RunTestNoThrow(Tests.GreaterInBound);
+        RunTestNoThrow(Tests.GreaterEqualInBound);
+        RunTestNoThrow(Tests.LessInBound);
+        RunTestNoThrow(Tests.EqualsInBound);
+        RunTestNoThrow(Tests.LessEqualsInBound);
+        RunTestNoThrow(Tests.EqualsInBound);
+        RunTestNoThrow(Tests.EqualsReversedInBound);
+        RunTestNoThrow(Tests.ZeroInBounds);  
+        RunTestNoThrow(Tests.CompareAgainstLong);
+        RunTestNoThrow(TestsEarlyReturn.GreaterInBound);
+        RunTestNoThrow(TestsEarlyReturn.GreaterEqualInBound);
+        RunTestNoThrow(TestsEarlyReturn.LessInBound);
+        RunTestNoThrow(TestsEarlyReturn.LessEqualsInBound);
+        RunTestNoThrow(TestsEarlyReturn.NotEqualsInBound);
+        RunTestNoThrow(TestsEarlyReturn.ZeroInBounds);
+
+        Tests.ModInBounds(arr, 11);
+        try { Tests.ModOutOfBounds(arr, 6); returnCode--; } catch {}
+
+        arr = new int[0];
+        RunTestThrows(Tests.ZeroOutOfBounds);
+        RunTestThrows(TestsEarlyReturn.ZeroOutOfBounds);
+
+        arr = new int[1];
+        RunTestThrows(Tests.OneOutOfBounds);
+        RunTestThrows(Tests.OneEqualsOutOfBounds);
+
+        return returnCode;
+    }
+
+    private static void RunTestThrows(Action<int[]> action)
+    {
+        try
+        {
+            action(arr);
+            Console.WriteLine("failed " + action.Method.Name);
+            returnCode--;
+        }
+        catch (Exception)
+        {
+
+        }
+    }
+
+    private static void RunTestNoThrow(Action<int[]> action)
+    {
+        try
+        {
+            action(arr);
+        }
+        catch (Exception)
+        {
+            Console.WriteLine("failed " + action.Method.Name);
+            returnCode--;
+        }
+    }
+}
+
+public static class Tests
+{
+    public static void GreaterInBound(int[] arr)
```

Comment:
Do we already handle all of these cases?
If not can you add a comment saying that we either
1) always remove the array bounds check
2) we don't currently remove the bonds check

Reply:
This handles all of these cases. More specifically, of the ones listed here: https://github.com/dotnet/runtime/issues/11623#issuecomment-455777587 we will now handle all of the signed cases and also equals for the unsigned case.